diff --git a/config/common_base b/config/common_base
index 2e0c72598..37192e2a9 100644
--- a/config/common_base
+++ b/config/common_base
@@ -84,18 +84,18 @@ CONFIG_RTE_MAX_HEAPS=32
 CONFIG_RTE_MAX_MEMSEG_LISTS=64
 # each memseg list will be limited to either RTE_MAX_MEMSEG_PER_LIST pages
 # or RTE_MAX_MEM_MB_PER_LIST megabytes worth of memory, whichever is smaller
-CONFIG_RTE_MAX_MEMSEG_PER_LIST=8192
-CONFIG_RTE_MAX_MEM_MB_PER_LIST=32768
+CONFIG_RTE_MAX_MEMSEG_PER_LIST=32784
+CONFIG_RTE_MAX_MEM_MB_PER_LIST=131072
 # a "type" is a combination of page size and NUMA node. total number of memseg
 # lists per type will be limited to either RTE_MAX_MEMSEG_PER_TYPE pages (split
 # over multiple lists of RTE_MAX_MEMSEG_PER_LIST pages), or
 # RTE_MAX_MEM_MB_PER_TYPE megabytes of memory (split over multiple lists of
 # RTE_MAX_MEM_MB_PER_LIST), whichever is smaller
-CONFIG_RTE_MAX_MEMSEG_PER_TYPE=32768
-CONFIG_RTE_MAX_MEM_MB_PER_TYPE=131072
+CONFIG_RTE_MAX_MEMSEG_PER_TYPE=131072
+CONFIG_RTE_MAX_MEM_MB_PER_TYPE=524288
 # global maximum usable amount of VA, in megabytes
 CONFIG_RTE_MAX_MEM_MB=524288
-CONFIG_RTE_MAX_MEMZONE=2560
+CONFIG_RTE_MAX_MEMZONE=10240
 CONFIG_RTE_MAX_TAILQ=32
 CONFIG_RTE_ENABLE_ASSERT=n
 CONFIG_RTE_LOG_DP_LEVEL=RTE_LOG_INFO
